net/coova-chilli/Makefile

```diff
@@ -75,7 +77,7 @@ endef
 
 DISABLE_NLS=
 
-TARGET_CFLAGS += $(FPIC)
+TARGET_CFLAGS += $(FPIC) -Wno-error
```

Comment:
What warnings are you hitting?  I'm compiling with 7.4.0, and aside from a `const` getting discarded in `src/ssl.c`, which you can avoid by casting to `(void *)` (see my proposed patch in the review comments), it compiles fine.

Reply:
Address of packet struct error. GCC8 and up.

http://downloads.openwrt.org/snapshots/faillogs/arc_arc700/packages/coova-chilli/compile.txt

Reply:
Oh this is a different one. I get -Werror=address-of-packed on GCC9.

Reply:
I can't give you 100% guarantee, but I am 99.99999999999999999% sure it is in fact a bug.  The [commit](https://github.com/coova/coova-chilli/commit/a0a0e65adf829d58d24461345b641d41fdfb4c39#diff-db5376f39c9e516b8a94af8d280d3d14) that added the offending `if` statement doesn't even have a title, but [in the beginning of time](https://github.com/coova/coova-chilli/blob/c36bbc6b77010b8e9d6826ec47cc57fb9e43d121/src/redir.c#L1936), `redir_memcopy`  used to be a function, later changed to a macro.  Besides, the effect of the if clause applying only to the first statement would be to not generate a new challenge, but it would conflict with the rest of the macro (copying the challenge, and displaying the reset message, all based on a `challengetimeout` option.  coova/coova-chilli#444 has been open for a while with no answer.
Here's what I would do:
```diff
--- a/src/redir.c
+++ b/src/redir.c
@@ -3358,14 +3358,17 @@ int redir_main(struct redir_t *redir,
   }

 #define redir_memcopy(msgtype)                                          \
+ do {                                                                   \
   redir_challenge(challenge);                                           \
   redir_chartohex(challenge, hexchal, REDIR_MD5LEN);                    \
   msg.mtype = msgtype;                                                  \
   memcpy(conn.s_state.redir.uamchal, challenge, REDIR_MD5LEN);          \
-  if (_options.debug) syslog(LOG_DEBUG, "%s(%d): ---->>> resetting challenge: %s", __FUNCTION__, __LINE__, hexchal)
+  if (_options.debug) syslog(LOG_DEBUG, "%s(%d): ---->>> resetting challenge: %s", __FUNCTION__, __LINE__, hexchal) \
+ } while (0)

 #ifdef USING_IPC_UNIX
 #define redir_msg_send(msgopt)                                          \
+ do {                                                                   \
   msg.mdata.opt = msgopt;                                               \
   memcpy(&msg.mdata.address, address, sizeof(msg.mdata.address));       \
   memcpy(&msg.mdata.baddress, baddress, sizeof(msg.mdata.baddress));    \
@@ -3375,9 +3378,11 @@ int redir_main(struct redir_t *redir,
     syslog(LOG_ERR, "%s: write() failed! msgfd=%d type=%ld len=%d",     \
            strerror(errno), redir->msgfd, msg.mtype, (int)sizeof(msg.mdata)); \
     return redir_main_exit(&socket, forked, rreq);                      \
-  }
+  }                                                                     \
+ } while (0)
 #else
 #define redir_msg_send(msgopt)                                          \
+ do {                                                                   \
   msg.mdata.opt = msgopt;                                               \
   memcpy(&msg.mdata.address, address, sizeof(msg.mdata.address));       \
   memcpy(&msg.mdata.baddress, baddress, sizeof(msg.mdata.baddress));    \
@@ -3387,7 +3392,8 @@ int redir_main(struct redir_t *redir,
     syslog(LOG_ERR, "%s: msgsnd() failed! msgid=%d type=%ld len=%d",    \
            strerror(errno), redir->msgid, msg.mtype, (int)sizeof(msg.mdata)); \
     return redir_main_exit(&socket, forked, rreq);                      \
-  }
+  }                                                                     \
+ } while (0)
 #endif

   /*
```

Reply:
The AUR package has a different solution: https://aur.archlinux.org/cgit/aur.git/tree/4149a5ddfcc666ebbca3fcc9da308e7ff9e433b7.patch?h=coova-chilli-arch

Reply:
I added this patch. It exposed a bug in the macro (missing ; in that if statement).

Reply:
These are the GCC9 warnings btw: https://gist.github.com/neheb/930695e8b7b6e8f0e45d67f0a038ad33

I refuse to touch those.

Reply:
That does not change the code at all, and will not avoid the warning.
The offending `if` statement is [this](https://github.com/coova/coova-chilli/blob/4149a5ddfcc666ebbca3fcc9da308e7ff9e433b7/src/redir.c#L4037):
```C
if (_options.challengetimeout)
  redir_memcopy(REDIR_CHALLENGE);
```
which, as is, will be expanded to (indentations are on me):
```C
if (_options.challengetimeout)
  redir_challenge(challenge);
redir_chartohex(challenge, hexchal, REDIR_MD5LEN);
msg.mtype = msgtype;
memcpy(conn.s_state.redir.uamchal, challenge, REDIR_MD5LEN);
if (_options.debug)
  syslog(LOG_DEBUG, "%s(%d): ---->>> resetting challenge: %s", __FUNCTION__, __LINE__, hexchal);
```
Notice that the if statement guards only the first macro statement.  What is usually intended is this:
```C
if (_options.challengetimeout) {
  redir_challenge(challenge);
  redir_chartohex(challenge, hexchal, REDIR_MD5LEN);
  msg.mtype = msgtype;
  memcpy(conn.s_state.redir.uamchal, challenge, REDIR_MD5LEN);
  if (_options.debug)
    syslog(LOG_DEBUG, "%s(%d): ---->>> resetting challenge: %s", __FUNCTION__, __LINE__, hexchal);
}
```
My code actually turns it into this, with the same effect, but more resilient to _some_ abuse:
```C
if (_options.challengetimeout)
  do {
    redir_challenge(challenge);
    redir_chartohex(challenge, hexchal, REDIR_MD5LEN);
    msg.mtype = msgtype;
    memcpy(conn.s_state.redir.uamchal, challenge, REDIR_MD5LEN);
    if (_options.debug)
      syslog(LOG_DEBUG, "%s(%d): ---->>> resetting challenge: %s", __FUNCTION__, __LINE__, hexchal);
  } while (0);
```

Reply:
Ah, very subtle.

Reply:
> 
> 
> I added this patch. It exposed a bug in the macro (missing ; in that if statement).

The macro is fine without the `;`.  It is part of the code before the expansion.
This is easy to see
```C
#define x(a, b) foo(a,b)

int main() {
  int a, b;
  x(a, b);
}
```
where `x(a, b);` becomes `foo(a, b);`
But if you have multiple statements, it gets ugly, but it is right nonetheless.
```C
#define x(a, b) foo(a,b); \
    bar(a,b) 
int main() {
  int a, b;
  x(a, b);
}
```
The final semicolon is taken from the `x(a, b)`**;** line, thus the expansion becomes 
`foo(a, b);` this semicolon comes from the macro
`bar(a, b)`**;** this one comes from the original `x(a, b);` line.

Reply:
> 
> 
> These are the GCC9 warnings btw: https://gist.github.com/neheb/930695e8b7b6e8f0e45d67f0a038ad33
> 
> I refuse to touch those.

Agreed, but you can turn that specific warning off.  Don't need to worry about unrecognized ´-Wno-` argument (from the `info gcc`):
```
When an unrecognized warning option is requested (e.g.,
'-Wunknown-warning'), GCC emits a diagnostic stating that the option is
not recognized.  However, if the '-Wno-' form is used, the behavior is
slightly different: no diagnostic is produced for '-Wno-unknown-warning'
unless other diagnostics are being produced.  This allows the use of new
'-Wno-' options with old compilers, but if something goes wrong, the
compiler warns that an unrecognized option is present.
```

Reply:
Actually I had to add the semicolon as it was failing to compile after I added the do, while statements.

I guess I'll add -Wno-address-of-packed-member instead.

Reply:
Got it, my bad.  The `;` was missing before the closing bracket.